### PR TITLE
fix(challenge): urlsafe base64 encoding

### PR DESCRIPTION
**Pull request by matoous**
This fixes issue with wrong encoding being used for the challenge. According to the specs[^1] the challenge should be base64 url-safe encoded. Until now, the package used std encoding which uses slightly different set of characters. This was reported in the linked issue and we also encountered the same issue when testing our implementation of webauthn e2e using github.com/descope/virtualwebauthn.

The commit removes the custom type for challenge and instead passes around the challenge url-safe base64 encoded right from the point of creation. Because the string is fully valid it can be safely used in json as well as stored in redis or other possible storage implementations that might be used by the clients of the library.

I tried to keep the tests as close to the original as possible. In some cases that required transforming the challenge back to bytes considering the previously used padding to be part of the bytes and then re-encoding into url-safe base64 string.

[^1]: https://www.w3.org/TR/webauthn-2/#dom-collectedclientdata-challenge

Fix: https://github.com/duo-labs/webauthn/issues/128

### CODE REVIEW

**Comment by james-d-elliott** [comment]
Hey Matouš! Thanks for contributing! I am relatively certain this change concept is correct. Have you checked this with existing credentials and new credentials by chance? I will probably do so if not just to ensure there are no weird issues.

**Comment by matoous** [comment]
@james-d-elliott I only tested the whole thing using an integration test setup we (@ company) have with https://github.com/descope/virtualwebauthn. We are currently working on webauthn capabilities and don't yet have the front end ready so I sadly don't have better way to test this end to end in real browser scenario. On the other hand, the https://github.com/descope/virtualwebauthn is what complained about the non-user-safe encoding and now our tests pass so I am confident the changes fix the issue. Still, if you please have a way to test this in real scenario, please do 🙏

**Comment by matoous** [comment]
@james-d-elliott please let me know if I should do any further testing or if any changes are required.

**Comment by james-d-elliott** [comment]
Hey I definitely will in the next few days, works just been a bit busy. I had a few questions mostly.

**Comment by james-d-elliott** [comment]
Oh I think I see.. because of the JSON Marshal.

File: protocol/assertion_test.go [hunk 1]
```diff
@@ -151,7 +151,7 @@ func TestParsedCredentialAssertionData_Verify(t *testing.T) {
 		Raw                       CredentialAssertionResponse
 	}
 	type args struct {
-		storedChallenge    Challenge
+		storedChallenge    string
```
Comment (james-d-elliott) [inline protocol/assertion_test.go:154]:
```suggestion
		storedChallenge    URLEncodedBase64
```

File: protocol/assertion_test.go [hunk 2]
```diff
@@ -173,7 +173,7 @@ func TestParsedCredentialAssertionData_Verify(t *testing.T) {
 				Raw:                       tt.fields.Raw,
 			}
 
-			if err := p.Verify(tt.args.storedChallenge.String(), tt.args.relyingPartyID, tt.args.relyingPartyOrigin, "", tt.args.verifyUser, tt.args.credentialBytes); (err != nil) != tt.wantErr {
+			if err := p.Verify(tt.args.storedChallenge, tt.args.relyingPartyID, tt.args.relyingPartyOrigin, "", tt.args.verifyUser, tt.args.credentialBytes); (err != nil) != tt.wantErr {
```
Comment (james-d-elliott) [inline protocol/assertion_test.go:176]:
```suggestion
			if err := p.Verify(tt.args.storedChallenge.String(), tt.args.relyingPartyID, tt.args.relyingPartyOrigin, "", tt.args.verifyUser, tt.args.credentialBytes); (err != nil) != tt.wantErr {
```

File: protocol/attestation_test.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ func TestAttestationVerify(t *testing.T) {
 			pcc.Response = *parsedAttestationResponse
 
 			// Test Base Verification
-			err = pcc.Verify(options.Response.Challenge.String(), false, options.Response.RelyingParty.ID, options.Response.RelyingParty.Name)
+			err = pcc.Verify(options.Response.Challenge, false, options.Response.RelyingParty.ID, options.Response.RelyingParty.Name)
```
Comment (james-d-elliott) [inline protocol/attestation_test.go:33]:
```suggestion
			err = pcc.Verify(options.Response.Challenge.String(), false, options.Response.RelyingParty.ID, options.Response.RelyingParty.Name)
```

File: protocol/challenge.go [hunk 4]
```diff
@@ -8,20 +8,13 @@ import (
...
 
-// Challenge that should be signed and returned by the authenticator
-type Challenge URLEncodedBase64
-
-// Create a new challenge to be sent to the authenticator. The spec recommends using
-// at least 16 bytes with 100 bits of entropy. We use 32 bytes.
-func CreateChallenge() (Challenge, error) {
+// Create a new challenge that should be signed and returned by the authenticator.
+// The spec recommends using at least 16 bytes with 100 bits of entropy. We use 32 bytes.
+func CreateChallenge() (string, error) {
```
Comment (james-d-elliott) [inline protocol/challenge.go:13]:
```suggestion
func CreateChallenge() (URLEncodedBase64, error) {
```

File: protocol/challenge.go [hunk 5]
```diff
@@ -8,20 +8,13 @@ import (
...
 	if err != nil {
-		return nil, err
+		return "", err
 	}
-	return challenge, nil
-}
-
-func (c Challenge) String() string {
-	return base64.RawURLEncoding.EncodeToString(c)
+	return base64.RawURLEncoding.EncodeToString(challenge), nil
```
Comment (james-d-elliott) [inline protocol/challenge.go:19]:
```suggestion
	return URLEncodedBase64(challenge), nil
```

File: protocol/options_test.go [hunk 6]
```diff
@@ -22,7 +22,7 @@ func TestPublicKeyCredentialRequestOptions_GetAllowedCredentialIDs(t *testing.T)
 		{
 			"Correct Credential IDs",
 			fields{
-				Challenge: Challenge([]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}),
+				Challenge: string([]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}),
```
Comment (james-d-elliott) [inline protocol/options_test.go:25]:
```suggestion
				Challenge: URLEncodedBase64([]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}),
```

File: protocol/options_test.go [hunk 7]
```diff
@@ -7,7 +7,7 @@ import (
 
 func TestPublicKeyCredentialRequestOptions_GetAllowedCredentialIDs(t *testing.T) {
 	type fields struct {
-		Challenge          Challenge
+		Challenge          string
```
Comment (james-d-elliott) [inline protocol/options_test.go:10]:
```suggestion
		Challenge          URLEncodedBase64
```

File: webauthn/login.go [hunk 8]
```diff
@@ -64,7 +63,7 @@ func (webauthn *WebAuthn) beginLogin(userID []byte, allowedCredentials []protoco
 	}
 
 	sessionData := SessionData{
-		Challenge:            base64.RawURLEncoding.EncodeToString(challenge),
+		Challenge:            challenge,
```
Comment (james-d-elliott) [inline webauthn/login.go:66]:
```suggestion
		Challenge:            challenge.String(),
```

File: webauthn/registration.go [hunk 9]
```diff
@@ -57,7 +56,7 @@ func (webauthn *WebAuthn) BeginRegistration(user User, opts ...RegistrationOptio
 
 	response := protocol.CredentialCreation{Response: creationOptions}
 	newSessionData := SessionData{
-		Challenge:        base64.RawURLEncoding.EncodeToString(challenge),
+		Challenge:        challenge,
```
Comment (james-d-elliott) [inline webauthn/registration.go:59]:
```suggestion
		Challenge:        challenge.String(),
```

File: protocol/options.go [hunk 10]
```diff
@@ -30,7 +30,7 @@ type PublicKeyCredentialCreationOptions struct {
 // Its challenge member MUST be present, while its other members are OPTIONAL.
 // See §5.5. Options for Assertion Generation https://www.w3.org/TR/webauthn/#assertion-options
 type PublicKeyCredentialRequestOptions struct {
-	Challenge          Challenge                   `json:"challenge"`
+	Challenge          string                      `json:"challenge"`
 	Timeout            int                         `json:"timeout,omitempty"`
```
Comment (james-d-elliott) [inline protocol/options.go:34]:
```suggestion
	Challenge          URLEncodedBase64                 `json:"challenge"`
	Timeout            int                         `json:"timeout,omitempty"`
```

File: protocol/credential_test.go [hunk 11]
```diff
@@ -236,7 +236,7 @@ func TestParsedCredentialCreationData_Verify(t *testing.T) {
 				Response:                  tt.fields.Response,
 				Raw:                       tt.fields.Raw,
 			}
-			if err := pcc.Verify(tt.args.storedChallenge.String(), tt.args.verifyUser, tt.args.relyingPartyID, tt.args.relyingPartyOrigin); (err != nil) != tt.wantErr {
+			if err := pcc.Verify(tt.args.storedChallenge, tt.args.verifyUser, tt.args.relyingPartyID, tt.args.relyingPartyOrigin); (err != nil) != tt.wantErr {
```
Comment (james-d-elliott) [inline protocol/credential_test.go:239]:
```suggestion
			if err := pcc.Verify(tt.args.storedChallenge.String(), tt.args.verifyUser, tt.args.relyingPartyID, tt.args.relyingPartyOrigin); (err != nil) != tt.wantErr {
```

File: protocol/credential_test.go [hunk 12]
```diff
@@ -162,7 +162,7 @@ func TestParsedCredentialCreationData_Verify(t *testing.T) {
 		Raw                       CredentialCreationResponse
 	}
 	type args struct {
-		storedChallenge    Challenge
+		storedChallenge    string
```
Comment (james-d-elliott) [inline protocol/credential_test.go:165]:
```suggestion
		storedChallenge    URLEncodedBase64
```

File: protocol/client_test.go [hunk 13]
```diff
@@ -26,9 +23,9 @@ func TestVerifyCollectedClientData(t *testing.T) {
 	var storedChallenge = newChallenge
 
 	originURL, _ := url.Parse(ccd.Origin)
-	err = ccd.Verify(storedChallenge.String(), ccd.Type, FullyQualifiedOrigin(originURL))
+	err = ccd.Verify(storedChallenge, ccd.Type, FullyQualifiedOrigin(originURL))
```
Comment (james-d-elliott) [inline protocol/client_test.go:26]:
```suggestion
	err = ccd.Verify(storedChallenge.String(), ccd.Type, FullyQualifiedOrigin(originURL))
```

**Review by james-d-elliott — COMMENTED** [review]
I think this mostly looks good. The reason it's not encoding it correctly is because the `Challenge` type is defined as a `URLEncodedBase64` but the go compiler actually treats it as a `[]byte`. There are multiple ways to work around this, we can use a struct with the following signature:

```go
type Challenge struct {
	URLEncodedBase64
}
```

Or we can get rid of it entirely like you have.

Or we can replace all instances of it with `URLEncodedBase64` which may be an better change overall? My suggestions show (roughly) how to do this.

**Comment by matoous** [comment]
@james-d-elliott thanks for suggestions, applied and adjusted. It's way closer now to the initial implementation.

**Comment by james-d-elliott** [comment]
> @james-d-elliott thanks for suggestions, applied and adjusted. It's way closer now to the initial implementation.

That's what I was going for. Thanks a lot! I'll do a final pass today and merge this hopefully.

**Review by james-d-elliott — APPROVED** [review]
LGTM

**Comment by tobiaszheller** [comment]
Hey, just want to leave some information in case someone updates from v0.5.0 to v.0.6.0, this PR will cause breaking changes on server - client connection if your client expects base64 instead of base64url encoding. 
I think it was not mention explicitly in release notes (or maybe it's not needed on pre v1). 

Anyway, thanks for keeping it inline with spec now :)

**Comment by james-d-elliott** [comment]
I'm on the fence as to if this was the breaking change or the change which introduced this bug was the breaking change. I'm more than willing to make special note of it in the release notes but I'm inclined to decline the idea this is breaking (specifically from a semver standpoint) as anytime something we're not doing to spec we will have the same expectations in many instances; but the spec is the contract.

I'll do some research regarding #93 however.

**Pull request merged**